tools/wptrunner/wptrunner/testharnessreport.js

```diff
@@ -1,75 +1,6 @@
-class MessageQueue {
-  constructor() {
-    this.item_id = 0;
-    this._queue = [];
-  }
-
-  push(item) {
-    let cmd_id = this.item_id++;
-    item.id = cmd_id;
-    this._queue.push(item);
-    __wptrunner_process_next_event();
-    return cmd_id;
-  }
-
-  shift() {
-    return this._queue.shift();
-  }
-}
-
-window.__wptrunner_testdriver_callback = null;
-window.__wptrunner_message_queue = new MessageQueue();
-window.__wptrunner_url = null;
-
-window.__wptrunner_process_next_event = function() {
-  /* This function handles the next testdriver event. The presence of
-     window.testdriver_callback is used as a switch; when that function
-     is present we are able to handle the next event and when is is not
-     present we must wait. Therefore to drive the event processing, this
-     function must be called in two circumstances:
-       * Every time there is a new event that we may be able to handle
-       * Every time we set the callback function
-     This function unsets the callback, so no further testdriver actions
-     will be run until it is reset, which wptrunner does after it has
-     completed handling the current action.
-   */
-
-  if (!window.__wptrunner_testdriver_callback) {
-    return;
-  }
-  var data = window.__wptrunner_message_queue.shift();
-  if (!data) {
-    return;
-  }
-
-  var payload = undefined;
-
-  switch(data.type) {
-  case "complete":
-    var tests = data.tests;
-    var status = data.status;
-
-    var subtest_results = tests.map(function(x) {
-      return [x.name, x.status, x.message, x.stack];
-    });
-    payload = [status.status,
-               status.message,
-               status.stack,
-               subtest_results];
-    clearTimeout(window.__wptrunner_timer);
-    break;
-  case "action":
-    payload = data;
-    break;
-  default:
-    return;
-  }
-  var callback = window.__wptrunner_testdriver_callback;
-  window.__wptrunner_testdriver_callback = null;
-  callback([__wptrunner_url, data.type, payload]);
-};
-
 (function() {
+  window.__wptrunner_is_test_context = true;
```

Comment:
This is to say as long as testharnessreport.js is included, it is in test context? Assume reftests won't include this file, this won't be true for them?

Reply:
I still think testharnessreport-based detection is the best way. Injecting the main context flag via resume script won't work because it's not guaranteed to happen before a test script already starts issuing testdriver actions (e.g., before `load`). Therefore, `testdriver-extra.js` needs to rely on the existing test markup (or an effect by an earlier `<script>`).

Reply:
Ack. We can land this first, then have some more discussion in the next PR by involving some other ppl. I do not have a suggestion here.